### DISCUSSION TITLE
Is there a limit on how many users you can create in a RabbitMQ server?

### DISCUSSION THREAD

**Discussion by myasul** [rabbitmq/rabbitmq-server#7805] [Questions] [answered]
We are planning to create a RabbitMQ server for our thousands of IOT devices to connect to. Each IOT device will have a dedicated user. And we would like to know if there is a limit on how many users you can create and store in a RabbitMQ server.

**Comment by mkuratczyk** [comment, marked as answer]
How many are you thinking? Users are generally "cheap", since they don't have any runtime component. When testing some metadata operations like imports, I sometimes use a JSON file with 100k users: https://github.com/rabbitmq/sample-configs/tree/main/users
As with anything else, if you are planning to use software at scale, you should spend some time simulating the workload to make sure it will work as expected. You can easily generate/import files like the one above + simulate the traffic using https://perftest.rabbitmq.com/ or https://github.com/emqx/emqtt-bench (based on the protocol you want to use).

**Reply by myasul** [reply]
Thanks for the quick reply, @mkuratczyk ! Currently, we have about 10K devices, and I think we will hit 100K in a few years.

I'm glad that it's already been tested for 100K users! Thank you for providing ways to do simulation. This will help us a lot in testing!

**Comment by michaelklishin** [comment]
Tens of thousands of users won't be an issue. Hundreds of thousands of users per se should not be but
at that point you will [have to reduce per-connection memory footprint](https://rabbitmq.com/networking.html#tuning-for-large-number-of-connections), increase [stats/metrics emission interval](https://rabbitmq.com/runtime.html#cpu-reduce-idle-usage) from the default 5s
and deal with per-queue memory footprint if most connections are consumers.

**Reply by myasul** [reply]
Thank you for answering my question and providing ways to move forward when we scale!

**Comment by aheadricksg** [comment]
I'm about a year tardy to the party but thought I should mention this. Unless the http api has changed in the recent versions, which it doesn't look like, the `/api/users` http endpoint doesn't support pagination so all users are dumped out in one giant payload. The management UI uses this api and the users endpoint is used in the admin section of it. Once the number of users gets sufficiently high, the admin section starts to break and the latency also gets really high

**Reply by mkuratczyk** [reply]
Thanks for sharing. Sounds like a problem that can be solved if there's a need to support a really large number users. I guess we haven't heard about this since most deployments with many users use an external authentication backend.

**Reply by michaelklishin** [reply]
In my 13.5 years as a RabbitMQ contributor, I am yet to see a real world example where this would actually come up as a practical issue. 

For several years now, you don't have to use the management UI [for monitoring](https://www.rabbitmq.com/docs/prometheus). OAuth 2 or x.509-certificate based authentication may be a more pragmatic approach where the credential cardinality gets into many thousands.
